src/org/apache/hadoop/tools/Curl.php

```diff
@@ -194,13 +194,34 @@ private function _exec($options, $returnInfo = false)
         }
         if (isset($this->options['local_file_handler'])) {
             $fp = $this->options['local_file_handler'];
-            flock($fp,LOCK_EX);
+            flock($fp, LOCK_EX);
             $options[CURLOPT_WRITEFUNCTION] = function ($ch, $string) use ($fp) {
                 $length = fwrite($fp, $string);
+
                 return $length;
             };
             fflush($fp);
         }
+        // auto add content-length header
+        $has_content_length_header = false;
+        foreach ($options[CURLOPT_HTTPHEADER] as $header) {
+            if (stripos($header, 'content-length:') === 0) {
+                $has_content_length_header = true;
+                break;
+            }
+        }
+        if (!$has_content_length_header && !isset($options[CURLOPT_INFILE]) && !isset($options[CURLOPT_INFILESIZE])) {
+            $length = 0;
+            if (isset($options[CURLOPT_POSTFIELDS])) {
+                if (function_exists('mb_strlen')) {
+                    $length = mb_strlen($options[CURLOPT_POSTFIELDS]);
```

Comment:
The `Content-Length` HTTP header should be a pure binary character length of the body, aka raw bytes, not a human readable length. The former would be always calculated using `strlen`, and the latter would be `mb_strlen`. Let's use `strlen` always here. 

Source: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Length

Reply:
👍 sweet find!